Source/MLXNN/Upsample.swift

```diff
@@ -0,0 +1,220 @@
+                \(x.ndim)D was provided
+                """)
+        }
+
+        let scaleFactor = self.scaleFactor.asArray(dimensions: dimensions)
+
+        switch mode {
+        case .nearest:
+            return upsampleNearest(x, scale: scaleFactor)
+        case .linear(let alignCorners):
+            return upsampleLinear(x, scale: scaleFactor, alignCorners: alignCorners)
+        }
+    }
+}
+
+private func upsampleNearest(_ x: MLXArray, scale: [Float]) -> MLXArray {
+    let dimensions = x.ndim - 2
+    precondition(
+        dimensions == scale.count, "A scale needs to be provided for each spatial dimension")
+
+    // get a truncated version of the scales
+    let intScale = scale.map { Int($0) }
+    let intFloatScale = intScale.map { Float($0) }
+
+    if intFloatScale == scale {
+        // Int scale means we can simply expand-broadcast and reshape
+        var shape = x.shape
+        for d in 0 ..< dimensions {
+            shape.insert(1, at: 2 + 2 * d)
+        }
+        var x = x.reshaped(shape)
+
+        for d in 0 ..< dimensions {
+            shape[2 + 2 * d] = intScale[d]
+        }
+        x = broadcast(x, to: shape)
+
+        for d in 0 ..< dimensions {
+            shape[d + 1] *= shape[d + 2]
+            shape.remove(at: d + 2)
+        }
+        x = x.reshaped(shape)
+
+        return x
+    } else {
+        // Float scales
+        let N = x.shape.dropFirst().dropLast()
+
+        var indices: [MLXArrayIndex] = [0...]
+        for (i, (n, s)) in zip(N, scale).enumerated() {
+            indices.append(nearestIndices(dimension: n, scale: s, dim: i, ndim: dimensions))
+        }
+
+        return x[indices]
+    }
+}
+
+private typealias IndexWeight = (MLXArray, MLXArray)
+
+private func upsampleLinear(_ x: MLXArray, scale: [Float], alignCorners: Bool) -> MLXArray {
+    let dimensions = x.ndim - 2
+    precondition(
+        dimensions == scale.count, "A scale needs to be provided for each spatial dimension")
+
+    let N = x.shape.dropFirst().dropLast()
+
+    // compute the sampling grid
+    var indexWeights = [(IndexWeight, IndexWeight)]()
+    for (i, (n, s)) in zip(N, scale).enumerated() {
+        indexWeights.append(
+            linearIndices(
+                dimension: n, scale: s, alignCorners: alignCorners, dim: i, ndim: dimensions))
+    }
+
+    // sample and compute the weights
+    var samples = [MLXArray]()
+    var weights = [MLXArray]()
+    for indexWeight in product(pairs: indexWeights) {
+        let index = indexWeight.map { $0.0 }
+        let weight = indexWeight.map { $0.1 }
+        samples.append(x[[0...] + index])
+        weights.append(weight.dropFirst().reduce(weight[0], { $0 * $1 }))
+    }
+
+    // interpolate
+    return zip(weights.dropFirst(), samples.dropFirst())
+        .reduce(weights[0] * samples[0]) { result, pair in
+            result + (pair.0 * pair.1)
+        }
+}
+
+private func product(pairs: [(IndexWeight, IndexWeight)]) -> [[IndexWeight]] {
```

Comment:
It doesn't really matter as it is private to the file and fairly natural for combinations of pairs, but I feel that people will be confused by this admittedly smart piece of code...

Reply:
yeah, it might be a bit opaque -- I can add a little more explanation

Reply:
@angeloskath see how this reads -- people unfamiliar with bit operations may still have to read carefully but at least it now explains what & why :-)

Reply:
Looks great :-)